File: src/simple_ros/simple_ros/logging_example.py
import rclpy
from rclpy.node import Node
from rclpy.logging import LoggingSeverity
from std_msgs.msg import String


class LoggerUsage(Node):
    def __init__(self):
        super().__init__("Logger_usage_demo")
        self.create_timer(1, self.on_timer)
        self.pub = self.create_publisher(String, "send", 10)
        self.number = 0

    def on_timer(self):
        self.get_logger().log(
            "timer_callback_called_(print once)",
             LoggingSeverity.INFO, 
             once=True)
        msg = String()
        msg.data = f"Current count: {self.number}"
        self.get_logger().info(f"Publishing: {msg.data}")
        self.pub.publish(msg)
        self.get_logger().info(msg.data)
        
        # Check if the count is a divisor of twelve
        if self.debug_function_to_evaluate():
            self.get_logger().debug("Count divides info 12")
        
        # Check if the count is even
        if self.number % 2 == 0:
            self.get_logger().debug("Count is even!")
        
        # Reset count if it exceeds 15
        if self.number > 15:
            self.get_logger().warn("Resetting count to 0!")
            self.number = 0
        
        # Increment the count
        self.number += 1
        
    def debug_function_to_evaluate(self):
        return is_divide_of_twelve(self.number, self.get_logger())
        
def is_divide_of_twelve(val, logger):
    if val == 0:
        logger.error("Module divisor cannot be 0")
        return False
    return (12 % val) == 0
    
def main():
    rclpy.init()
    node = LoggerUsage()
    try:
        rclpy.spin(node)
    except KeyboardInterrupt:
        node.destroy_node()

if __name__ == "__main__":
    main()

File: src/simple_ros_cpp/src/logging_example.cpp
#include "rclcpp/rclcpp.hpp"
#include "std_msgs/msg/string.hpp"
#include <chrono>
#include <iostream>
#include <functional>

using namespace std;
using namespace std::chrono_literals;

// Function declaration
bool is_divisor_of_twelve(size_t val, rclcpp::Logger logger);

class LoggerUsage : public rclcpp::Node
{
public:
    LoggerUsage()
        : Node("logger_usage_demo"), _count(0)
    {
        _pub = create_publisher<std_msgs::msg::String>("message", 10);
        _timer = create_wall_timer(1s, std::bind(&LoggerUsage::pub_callback, this));

        debug_function_to_evaluate = std::bind(is_divisor_of_twelve, std::cref(_count), get_logger());

        // Create a one-shot timer
        _one_shot_timer = create_wall_timer(5s, [this]() {
            RCLCPP_INFO(get_logger(), "One-shot timer triggered");
            _one_shot_timer->cancel(); // Cancel after it's triggered
        });
    }

private:
    int _count;
    rclcpp::Publisher<std_msgs::msg::String>::SharedPtr _pub;
    rclcpp::TimerBase::SharedPtr _timer;
    rclcpp::TimerBase::SharedPtr _one_shot_timer;
    std::function<bool()> debug_function_to_evaluate;

    void pub_callback()
    {
        auto msg = std_msgs::msg::String();
        msg.data = "Current count: " + std::to_string(_count);
        
        RCLCPP_INFO(get_logger(), "Publishing: %s", msg.data.c_str());
        _pub->publish(msg);

        // Check if count is a divisor of twelve
        if (debug_function_to_evaluate())
        {
            RCLCPP_DEBUG(get_logger(), "Count is a divisor of 12");
        }

        // Check if count is even
        if (_count % 2 == 0)
        {
            RCLCPP_DEBUG(get_logger(), "Count is even!");
        }

        // Increment the count
        _count++;
        
        // Reset count if it exceeds 15
        if (_count > 15)
        {
            RCLCPP_WARN(get_logger(), "Resetting count to 0!");
            _count = 0;
        }
    }
};

bool is_divisor_of_twelve(size_t val, rclcpp::Logger logger)
{
    if (val == 0)
    {
        RCLCPP_ERROR(logger, "Module divisor cannot be 0");
        return false;
    }
    return (12 % val) == 0;
}

int main(int argc, char **argv)
{
    rclcpp::init(argc, argv);
    auto node = std::make_shared<LoggerUsage>();
    rclcpp::spin(node);
    rclcpp::shutdown();
    return 0;
}
